pkg/flavor/flavor.go

```diff
@@ -34,23 +41,230 @@ func (c Conflicts) String() string {
+
+	revisionPatch, err := json.Marshal(preferenceSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            revisionName,
+			Namespace:       vm.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(vm, virtv1.VirtualMachineGroupVersionKind)},
+		},
+		Data: runtime.RawExtension{Raw: revisionPatch},
+	}, nil
+
+}
+
+func (m *methods) createPreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	switch strings.ToLower(vm.Spec.Preference.Kind) {
+	case apiflavor.SingularPreferenceResourceName, apiflavor.PluralPreferenceResourceName:
+		preference, err := m.findPreference(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		revision, err := CreatePreferenceControllerRevision(vm, GetRevisionName(vm.Name, preference.Name, preference.UID, preference.Generation), &preference.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+	case apiflavor.ClusterSingularPreferenceResourceName, apiflavor.ClusterPluralPreferenceResourceName:
+		clusterPreference, err := m.findClusterPreference(vm)
+		if err != nil {
+			return nil, err
+		}
+
+		revision, err := CreatePreferenceControllerRevision(vm, GetRevisionName(vm.Name, clusterPreference.Name, clusterPreference.UID, clusterPreference.Generation), &clusterPreference.Spec)
+		if err != nil {
+			return nil, err
+		}
+		return revision, nil
+	default:
+		return nil, fmt.Errorf("got unexpected kind in PreferenceMatcher: %s", vm.Spec.Preference.Kind)
+	}
+}
+
+func (m *methods) storePreferenceRevision(vm *virtv1.VirtualMachine) (*appsv1.ControllerRevision, error) {
+
+	if vm.Spec.Preference == nil || len(vm.Spec.Preference.RevisionName) > 0 {
+		return nil, nil
+	}
+
+	preferenceRevision, err := m.createPreferenceRevision(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	preferenceRevision, err = m.clientset.AppsV1().ControllerRevisions(vm.Namespace).Create(context.Background(), preferenceRevision, metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return nil, err
+		}
+	}
+
+	vm.Spec.Preference.RevisionName = preferenceRevision.Name
+
+	return preferenceRevision, nil
+
+}
+
+func GenerateRevisionNamePatch(flavorRevision, preferenceRevision *appsv1.ControllerRevision) ([]byte, error) {
+
+	var flavorRevisionNamePatch, preferenceRevisionPatch utiltypes.PatchOperation
+
+	if flavorRevision != nil {
+		flavorRevisionNamePatch = utiltypes.PatchOperation{
+			Op:    utiltypes.PatchReplaceOp,
+			Path:  "/spec/flavor/revisionName",
+			Value: flavorRevision.Name,
+		}
+	}
+
+	if preferenceRevision != nil {
+		preferenceRevisionPatch = utiltypes.PatchOperation{
+			Op:    utiltypes.PatchReplaceOp,
+			Path:  "/spec/preference/revisionName",
+			Value: preferenceRevision.Name,
+		}
+	}
```

Comment:
should this be a `PatchAddOp` in the case that no preference or flavor revision is already set? 

Also, i think we need a `test` operation in the patch for safety in order to ensure we aren't stomping on another change. theoretically someone could set the revisionName themselves and if the timing was just right, the controller would stomp on the user's change. The test ensures we only mutate the revisionName if the existing value is what we think it should be. Take a look at other Patch operations to see what I mean about the `test` operation.

Reply:
I've tried to add the following test op but it's causing panics in kube-api, is there another way of expressing `null`?

```go
utiltypes.PatchOperation{
    Op:    utiltypes.PatchTestOp,
    Path:  "/spec/flavor/revisionName",
    Value: nil,
},
```
```
E0624 10:57:03.427737       1 runtime.go:76] Observed a panic: runtime error: invalid memory address or nil pointer dereference
goroutine 1874754 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest.func1.1()
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:105 +0xaf
panic({0x43124a0, 0x810bce0})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
k8s.io/kubernetes/vendor/github.com/evanphx/json-patch.Patch.test({0xc0081a6000, 0x47e0e20, 0xc006728b58}, 0xc01a7d4e78, 0xc02648b930)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/evanphx/json-patch/patch.go:691 +0x4aa
k8s.io/kubernetes/vendor/github.com/evanphx/json-patch.Patch.ApplyIndent({0xc0012d3300, 0x31, 0x4}, {0xc0081a6000, 0xefb, 0xefb}, {0x0, 0x0})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/evanphx/json-patch/patch.go:819 +0x24a
k8s.io/kubernetes/vendor/github.com/evanphx/json-patch.Patch.Apply(...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/evanphx/json-patch/patch.go:781
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*jsonPatcher).applyJSPatch(0xc00df75580, {0xc0081a6000, 0xefb, 0xefb})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:399 +0x867
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*jsonPatcher).applyPatchToCurrentObject(0xc00df75580, {0x54f7a38, 0xc01ddf9d40}, {0x54dc800, 0xc02648b8f8})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:321 +0xaa
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).applyPatch(0xc0184fab40, {0x54f7a38, 0xc01ddf9d40}, {0x4c3ce80, 0x4c71af1}, {0x54dc800, 0xc02648b8f8})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:563 +0xd3
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest.(*defaultUpdatedObjectInfo).UpdatedObject(0x8180198, {0x54f7a38, 0xc01ddf9d40}, {0x54dc800, 0xc02648b8f8})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/update.go:229 +0xd0
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update.func1({0x54dc800, 0xc02648b8f8}, {0xc0a583ffd975637e, 0x8fe93913d0b})
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:529 +0x1f9
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).updateState(0xc007b18900, 0xc02323de80, 0x3e)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:945 +0x3e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).GuaranteedUpdate(0xc007b18900, {0x54f7a38, 0xc01ddf9d40}, {0xc02323dd40, 0x35}, {0x54dc800, 0xc02648b8f0}, 0x1, 0x40d0b4, 0xc01b4b9130, ...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:365 +0x56e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).GuaranteedUpdate(0xc00a91ac60, {0x54f7a38, 0xc01ddf9d40}, {0xc02323dd40, 0x35}, {0x54dc800, 0xc02648b8f0}, 0xa8, 0x4b44120, 0xc01b4b9130, ...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:760 +0x1b5
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*DryRunnableStorage).GuaranteedUpdate(0x4c78de0, {0x54f7a38, 0xc01ddf9d40}, {0xc02323dd40, 0x4c68fa8}, {0x54dc800, 0xc02648b8f0}, 0xf2, 0xc00a4a8e68, 0xc01b4b9130, ...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/dryrun.go:101 +0x1c7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update(0xc00f612500, {0x54f7a38, 0xc01ddf9d40}, {0xc0066a8f88, 0xb}, {0x54dd368, 0xc0053633e0}, 0xc004d31090, 0xc022cab640, 0x0, ...)
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:517 +0x508
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).patchResource.func2()
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:662 +0xa7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).patchResource.func3()
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:668 +0x38
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest.func1()
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:117 +0x8f
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest
        /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:92 +0xe5

```

FWIW null does work manually with kubectl:

```sh
# ./cluster-up/kubectl.sh get vms/vm-cirros-csmall -o json | jq .spec.flavor
selecting docker as container runtime
{
  "kind": "VirtualMachineFlavor",
  "name": "vmf-csmall"
}
./cluster-up/kubectl.sh patch vms/vm-cirros-csmall --type='json' -p='[{"op": "test", "path": "/spec/flavor/revisionName", "value": null}]'
selecting docker as container runtime
virtualmachine.kubevirt.io/vm-cirros-csmall patched (no change)
```

Reply:
Maybe the reason is that the `PatchOperation` struct defines `Value` as `omitempty`:
https://github.com/kubevirt/kubevirt/blob/b541aad46a08e8493823a41164714c447b6463e4/pkg/util/types/patch.go#L27-L31

So when it is `nil`, `json.Marshal()` function ignores it and creates json without it:
```json
{
  "op": "test",
  "path": "/spec/flavor/revisionName"
}
```

Maybe that causes the panic?